### PR TITLE
Lower MIN/MAX/ISHFTC dynamic optional arguments

### PR DESCRIPTION
**Pull request by jeanPerier**
MIN/MAX/ISHFTC cases cannot fit in the default handling of dynamically
optional arguments because what needs to be done when the argument
is absent in lowering is not simply using a zero value or making a
nullptr.

These intrinsics need custom handling to delay the argument evaluation until the
implementation of the intrinsic code. In the usual framework, all intrinsics
arguments are lowered to ExtendedValue and passed to genIntrinsicCall(xxx).
This does not work here because we cannot pre-evaluate optional arguments to ExtendedValue before
calling the helper generating the intrinsic code.

In this patch, a callback is passed to generate the argument values 
to the intrinsic implementation so that the intrinsic lowering
code can generate the optional argument reference inside regions where
it tested that the argument must be present.

To avoid bloating `ConvertExpr.cpp`, this is done in a new
`CustomIntrinsicCall.cpp` file. This is not done in IntrinsicCall.cpp to keep
this last file independent from lib/Evaluate.
The new utility API is composed of three functions:
- One function to tell if an intrinsic call requires custom handling
- One function to prepare the call before any potential elemental loop nest (analyze
argument presence, prepare lambdas or arguments).
- One function to lower the call given two lambdas: one to access argument at index `i`,
  and one to query about the presence of argument at index `i`.

This may sounds complex, but it allows a single implementation of the intrinsics for the scalar/array case, while allowing to move the code outside of `ConvertExpr.cpp` and abstracting which intrinsics exactly must be handled by this utility.

### CODE REVIEW

**Comment by psteinfeld** [comment]
Note that the spelling of this branch is a little confusing.  "Shift" is abbreviated as "isfhtc" rather than "ishftc".  This baffled me for a few minutes.

File: flang/include/flang/Lower/CustomIntrinsicCall.h [hunk 1]
```diff
@@ -0,0 +1,99 @@
...
+    const Fortran::evaluate::ProcedureRef &procRef,
+    const Fortran::evaluate::SpecificIntrinsic &intrinsic,
+    AbstractConverter &converter);
+
+/// Type of callback to be provided to prepare the arguments fetching from an
+/// actual argument expression.
+using OperandPrepare = std::function<void(const Fortran::lower::SomeExpr &)>;
+
+/// Type of the callback to inquire about an argument presence, once the call
+/// prepartion was done. An absent optional means the argument is statically
```
Comment (psteinfeld) [inline flang/include/flang/Lower/CustomIntrinsicCall.h:53]: Should be "preparation" rather than "prepartion".

File: flang/include/flang/Lower/CustomIntrinsicCall.h [hunk 2]
```diff
@@ -0,0 +1,99 @@
...
+/// element value.
+using OperandGetter = std::function<fir::ExtendedValue(std::size_t)>;
+
+/// Given a callback \p prepareOptionalArgument to prepare optional
+/// arguments and a callback \p prepareOtherArgument to prepare non-optional
+/// arguments prepare the intrinsic arguments calls.
+/// It is up to the caller to decide what argument preparation means,
+/// the only contract is that it should later allow the caller to provide
+/// callbacks to generate argument reference given an argument index without
+/// any further knowledge of the argument. The function simply visit
```
Comment (psteinfeld) [inline flang/include/flang/Lower/CustomIntrinsicCall.h:73]: Should be "visits" rather than "visit".

File: flang/lib/Lower/CustomIntrinsicCall.cpp [hunk 3]
```diff
@@ -0,0 +1,261 @@
...
+#include "flang/Evaluate/fold.h"
+#include "flang/Evaluate/tools.h"
+#include "flang/Lower/IntrinsicCall.h"
+#include "flang/Lower/Todo.h"
+
+// CC genMinMaxWithOptionalArgs(const Fortran::evaluate::ProcedureRef &procRef,
+// const Fortran::evaluate::SpecificIntrinsic &intrinsic,
+// llvm::Optional<mlir::Type> retTy) {
+//   };
+// }
```
Comment (psteinfeld) [inline flang/lib/Lower/CustomIntrinsicCall.cpp:24]: Did you mean for this to be here?

**Review by psteinfeld — COMMENTED** [review]
I don't understand the code very well, so you should get an approval from someone who does.

I found a few nits, and I've confirmed that everything builds and tests correctly.

**Review by vdonaldson — COMMENTED** [review]
Looks ok to me

**Comment by schweitzpgi** [comment]
I'm looking at the code, but my initial question is fairly broad. Does it make sense to push any of the custom handling into the runtime C++ code here? Maybe different entry points, etc.

More abstractly: what is the trade-off between inline custom code templates for these calls and trying to push the customization into the runtime?

File: flang/lib/Lower/ConvertExpr.cpp [hunk 4]
```diff
@@ -2001,14 +1949,41 @@ class ScalarExprLowering {
     llvm::SmallVector<ExtValue> operands;
 
     llvm::StringRef name = intrinsic.name;
+    mlir::Location loc = getLoc();
+    if (Fortran::lower::intrinsicRequiresCustomOptionalHandling(
+            procRef, intrinsic, converter)) {
+      using ExvAndPresence = std::pair<ExtValue, llvm::Optional<mlir::Value>>;
+      llvm::SmallVector<ExvAndPresence, 4> operands;
```
Comment (schweitzpgi) [inline flang/lib/Lower/ConvertExpr.cpp:1956]: I assume the `4` is required by the `SmallVector` implementation.

File: flang/lib/Lower/ConvertExpr.cpp [hunk 5]
```diff
@@ -2001,14 +1949,41 @@ class ScalarExprLowering {
...
+        operands.emplace_back(optionalArg, isPresent);
+      };
+      auto prepareOtherArg = [&](const Fortran::lower::SomeExpr &expr) {
+        operands.emplace_back(genval(expr), llvm::None);
+      };
+      Fortran::lower::prepareCustomIntrinsicArgument(
+          procRef, intrinsic, resultType, prepareOptionalArg, prepareOtherArg,
+          converter);
+
+      llvm::StringRef name = intrinsic.name;
```
Comment (schweitzpgi) [inline flang/lib/Lower/ConvertExpr.cpp:1970]: Is this a duplicate of line 1951?

File: flang/lib/Lower/ConvertExpr.cpp [hunk 6]
```diff
@@ -2001,14 +1949,41 @@ class ScalarExprLowering {
...
+      auto getArgument = [&](std::size_t i) -> ExtValue {
+        if (fir::conformsWithPassByRef(
+                fir::getBase(operands[i].first).getType()))
+          return ::genLoad(builder, loc, operands[i].first);
```
Comment (schweitzpgi) [inline flang/lib/Lower/ConvertExpr.cpp:1974]: `genLoad` is in the global namespace? That seems like an oversight, no?

File: flang/lib/Lower/ConvertExpr.cpp [hunk 7]
```diff
@@ -4375,13 +4349,49 @@ class ArrayExprLowering {
...
-    if (isMinOrMaxWithDynamicallyOptionalArg(name, procRef,
-                                             converter.getFoldingContext()) ||
-        isIshftcWithDynamicallyOptionalArg(name, procRef,
-                                           converter.getFoldingContext()))
-      TODO(getLoc(),
-           "special handling of dynamically optional intrinsic argument");
     mlir::Location loc = getLoc();
+    if (Fortran::lower::intrinsicRequiresCustomOptionalHandling(
+            procRef, intrinsic, converter)) {
+      using CcPairT = std::pair<CC, llvm::Optional<mlir::Value>>;
```
Comment (schweitzpgi) [inline flang/lib/Lower/ConvertExpr.cpp:4355]: Just noting that `mlir::Value` has an existence property all by itself in that it can be a value or `nullptr`. Is the `Optional` wrapper needed?

File: flang/lib/Lower/ConvertExpr.cpp [hunk 8]
```diff
@@ -5628,7 +5622,34 @@ class ArrayExprLowering {
...
+      mlir::Value isPresent = genActualIsPresentTest(builder, loc, optionalArg);
+      mlir::Value elementValue =
+          fir::getBase(genOptionalValue(builder, loc, optionalArg, isPresent));
+      return [=](IterSpace iters) -> ExtValue { return elementValue; };
+    }
 
+    CC cc;
+    mlir::Value isPresent;
+    mlir::Type eleType;
+    std::tie(cc, isPresent, eleType) = genOptionalArrayFetch(expr);
```
Comment (schweitzpgi) [inline flang/lib/Lower/ConvertExpr.cpp:5652]: Does the `auto [a, b, c]` trick not work here?

File: flang/test/Lower/intrinsic-procedures/max.f90 [hunk 9]
```diff
@@ -0,0 +1,139 @@
...
+! CHECK:  fir.call @_FortranAioOutputInteger32(%{{.*}}, %[[VAL_13]]) : (!fir.ref<i8>, i32) -> i1
+end subroutine 
+
+! CHECK-LABEL: func @_QMmax_testPdynamic_optional_weird(
+! CHECK-SAME:  %[[VAL_0:.*]]: !fir.ref<i32> {fir.bindc_name = "a"},
+! CHECK-SAME:  %[[VAL_1:.*]]: !fir.ref<i32> {fir.bindc_name = "b"},
+! CHECK-SAME:  %[[VAL_2:.*]]: !fir.ref<i32> {fir.bindc_name = "c", fir.optional},
+! CHECK-SAME:  %[[VAL_3:.*]]: !fir.ref<i32> {fir.bindc_name = "d"},
+! CHECK-SAME:  %[[VAL_4:.*]]: !fir.ref<i32> {fir.bindc_name = "e", fir.optional}) {
+subroutine dynamic_optional_weird(a, b, c, d, e)
```
Comment (schweitzpgi) [inline flang/test/Lower/intrinsic-procedures/max.f90:95]: Good name. :)

File: flang/lib/Lower/ConvertExpr.cpp (hunk 8 above)
Comment (jeanPerier) [reply flang/lib/Lower/ConvertExpr.cpp:5652]: When using `auto [a, b, c]`, I do not know how to use `a`, `b`, `c` inside a lambda without getting a compiler error telling bind names cannot be captured.

File: flang/lib/Lower/CustomIntrinsicCall.cpp (hunk 3 above)
Comment (jeanPerier) [reply flang/lib/Lower/CustomIntrinsicCall.cpp:24]: No...

**Comment by jeanPerier** [comment]
> I'm looking at the code, but my initial question is fairly broad. Does it make sense to push any of the custom handling into the runtime C++ code here? Maybe different entry points, etc.
> 
> More abstractly: what is the trade-off between inline custom code templates for these calls and trying to push the customization into the runtime?

Yes, for some intrinsics like SYSTEM_CLOCK/RANDOM_SEED, I think custom runtime makes the most sense. The TODOs are in the added utility because it is the best place to catch this until runtime entries are added.

- For MIN/MAX, the runtime interface would not be trivial because of the variadic aspects (there is no standard defined maximum of arguments). So some level of custom lowering would still be required to generate those calls. It is not clear the complexity would be much less than the one to generate the inlined code. For numerical array min/max, having the branch generated in inlined code may also leave more slack for the optimizer to try to unswitch the code.

- For ISHFTC, the implementation was nearly free once the utility was done for MIN and MAX, and we already have a tested implementation of inlined ISHFTC (without dynamic optional argument aspects) that is tapped in directly once the optional aspects is dealt with. Dealing with it in the runtime would require making a completely new implementation  of it (with all the different versions per KIND) only for an edge case (low test coverage).

**Comment by jeanPerier** [comment]
Ping

**Review by schweitzpgi — APPROVED** [review]
